File: src/ai-handler.js
/**
 * AI Chat Handler - supports OpenAI and Anthropic API formats
 */
const https = require('https');

// Supported API formats
const API_FORMAT = {
  OPENAI: 'openai',
  ANTHROPIC: 'anthropic',
};

/**
 * Main AI chat handler
 */
async function handleAIChat(event, body) {
  const path = event.rawPath || event.requestContext?.http?.path || '/';
  const method = event.requestContext?.http?.method || event.httpMethod || 'POST';

  // Health check
  if (path === '/health' || path === '/') {
    if (method === 'GET') {
      return buildResponse(200, {
        status: 'ok',
        service: 'alexa-ai-lambda',
        endpoints: {
          openai: '/v1/chat/completions',
          anthropic: '/v1/messages',
          alexa: '/ (POST with Alexa request body)',
        },
      });
    }
  }

  if (method !== 'POST') {
    return buildResponse(405, { error: { message: 'Method not allowed', type: 'invalid_request_error' } });
  }

  // Route by path
  if (path === '/v1/chat/completions') {
    return await handleOpenAIFormat(body);
  } else if (path === '/v1/messages') {
    return await handleAnthropicFormat(body);
  }

  // Auto-detect format from body
  const format = detectFormat(body);
  if (format === API_FORMAT.OPENAI) {
    return await handleOpenAIFormat(body);
  } else if (format === API_FORMAT.ANTHROPIC) {
    return await handleAnthropicFormat(body);
  }

  return buildResponse(400, {
    error: {
      message: 'Could not determine API format. Use /v1/chat/completions for OpenAI or /v1/messages for Anthropic format.',
      type: 'invalid_request_error',
    },
  });
}

/**
 * Detect API format from request body
 */
function detectFormat(body) {
  if (!body) return null;
  // OpenAI format has 'messages' array with role/content objects
  if (body.messages && body.model && !body.max_tokens && !body.system) {
    return API_FORMAT.OPENAI;
  }
  // Anthropic format has 'messages' + 'max_tokens' and optional 'system'
  if (body.messages && body.max_tokens !== undefined) {
    return API_FORMAT.ANTHROPIC;
  }
  // Default: try OpenAI
  if (body.messages) return API_FORMAT.OPENAI;
  return null;
}

/**
 * Handle OpenAI Chat Completion format
 * POST /v1/chat/completions
 */
async function handleOpenAIFormat(body) {
  const { model, messages, temperature, max_tokens, stream } = body;

  if (!messages || !Array.isArray(messages)) {
    return buildResponse(400, {
      error: { message: 'messages is required and must be an array', type: 'invalid_request_error' },
    });
  }

  if (stream) {
    return buildResponse(400, {
      error: { message: 'Streaming is not supported in Lambda', type: 'invalid_request_error' },
    });
  }

  // Determine backend
  const backendModel = model || process.env.DEFAULT_MODEL || 'claude-sonnet-4-20250514';
  const backend = getBackendForModel(backendModel);

  let result;
  if (backend === 'anthropic') {
    // Convert OpenAI format to Anthropic and call
    const anthropicMessages = [];
    let systemPrompt = '';

    for (const msg of messages) {
      if (msg.role === 'system') {
        systemPrompt += (systemPrompt ? '\n' : '') + msg.content;
      } else {
        anthropicMessages.push({ role: msg.role, content: msg.content });
      }
    }

    result = await callAnthropicAPI({
      model: backendModel,
      messages: anthropicMessages,
      system: systemPrompt || undefined,
      max_tokens: max_tokens || 1024,
      temperature: temperature,
    });

    // Convert Anthropic response to OpenAI format
    return buildResponse(200, {
      id: `chatcmpl-${result.id || Date.now()}`,
      object: 'chat.completion',
      created: Math.floor(Date.now() / 1000),
      model: result.model,
      choices: [
        {
          index: 0,
          message: {
            role: 'assistant',
            content: result.content?.[0]?.text || '',
          },
          finish_reason: mapStopReason(result.stop_reason),
        },
      ],
      usage: {
        prompt_tokens: result.usage?.input_tokens || 0,
        completion_tokens: result.usage?.output_tokens || 0,
        total_tokens: (result.usage?.input_tokens || 0) + (result.usage?.output_tokens || 0),
      },
    });
  } else {
    // Call OpenAI API directly
    result = await callOpenAIAPI({
      model: backendModel,
      messages,
      temperature,
      max_tokens,
    });
    return buildResponse(200, result);
  }
}

/**
 * Handle Anthropic Messages format
 * POST /v1/messages
 */
async function handleAnthropicFormat(body) {
  const { model, messages, system, max_tokens, temperature, stream } = body;

  if (!messages || !Array.isArray(messages)) {
    return buildResponse(400, {
      error: {
        type: 'invalid_request_error',
        message: 'messages is required and must be an array',
      },
    });
  }

  if (stream) {
    return buildResponse(400, {
      error: {
        type: 'invalid_request_error',
        message: 'Streaming is not supported in Lambda',
      },
    });
  }

  const backendModel = model || process.env.DEFAULT_MODEL || 'claude-sonnet-4-20250514';
  const backend = getBackendForModel(backendModel);

  if (backend === 'openai') {
    // Convert Anthropic format to OpenAI and call
    const openaiMessages = [];
    if (system) {
      openaiMessages.push({ role: 'system', content: system });
    }
    for (const msg of messages) {
      openaiMessages.push({
        role: msg.role,
        content: typeof msg.content === 'string' ? msg.content : msg.content.map((c) => c.text).join(''),
      });
    }

    const result = await callOpenAIAPI({
      model: backendModel,
      messages: openaiMessages,
      temperature,
      max_tokens: max_tokens || 1024,
    });

    // Convert OpenAI response to Anthropic format
    const choice = result.choices?.[0];
    return buildResponse(200, {
      id: result.id || `msg_${Date.now()}`,
      type: 'message',
      role: 'assistant',
      content: [{ type: 'text', text: choice?.message?.content || '' }],
      model: result.model,
      stop_reason: mapFinishReason(choice?.finish_reason),
      usage: {
        input_tokens: result.usage?.prompt_tokens || 0,
        output_tokens: result.usage?.completion_tokens || 0,
      },
    });
  } else {
    // Call Anthropic API directly
    const result = await callAnthropicAPI({
      model: backendModel,
      messages,
      system,
      max_tokens: max_tokens || 1024,
      temperature,
    });
    return buildResponse(200, result);
  }
}

/**
 * Determine backend based on model name
 */
function getBackendForModel(model) {
  if (!model) return 'anthropic';
  const lower = model.toLowerCase();
  if (lower.startsWith('gpt-') || lower.startsWith('o1') || lower.startsWith('o3')) {
    return 'openai';
  }
  return 'anthropic';
}

/**
 * Call Anthropic Messages API
 */
async function callAnthropicAPI(params) {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    throw new Error('ANTHROPIC_API_KEY environment variable is not set');
  }

  const payload = JSON.stringify({
    model: params.model,
    messages: params.messages,
    system: params.system,
    max_tokens: params.max_tokens || 1024,
    temperature: params.temperature,
  });

  return await httpRequest({
    hostname: 'api.anthropic.com',
    path: '/v1/messages',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
    },
    body: payload,
  });
}

/**
 * Call OpenAI Chat Completions API
 */
async function callOpenAIAPI(params) {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    throw new Error('OPENAI_API_KEY environment variable is not set');
  }

  const payload = JSON.stringify({
    model: params.model,
    messages: params.messages,
    temperature: params.temperature,
    max_tokens: params.max_tokens,
  });

  return await httpRequest({
    hostname: 'api.openai.com',
    path: '/v1/chat/completions',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: payload,
  });
}

/**
 * Generic HTTPS request helper
 */
function httpRequest({ hostname, path, method, headers, body }) {
  return new Promise((resolve, reject) => {
    const req = https.request({ hostname, path, method, headers }, (res) => {
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          if (res.statusCode >= 400) {
            const err = new Error(parsed.error?.message || `API returned ${res.statusCode}`);
            err.statusCode = res.statusCode;
            err.response = parsed;
            reject(err);
          } else {
            resolve(parsed);
          }
        } catch (e) {
          reject(new Error(`Failed to parse API response: ${data.substring(0, 200)}`));
        }
      });
    });
    req.on('error', reject);
    req.setTimeout(30000, () => {
      req.destroy();
      reject(new Error('API request timeout'));
    });
    if (body) req.write(body);
    req.end();
  });
}

/**
 * Map Anthropic stop_reason to OpenAI finish_reason
 */
function mapStopReason(stopReason) {
  const map = { end_turn: 'stop', max_tokens: 'length', stop_sequence: 'stop' };
  return map[stopReason] || 'stop';
}

/**
 * Map OpenAI finish_reason to Anthropic stop_reason
 */
function mapFinishReason(finishReason) {
  const map = { stop: 'end_turn', length: 'max_tokens', content_filter: 'end_turn' };
  return map[finishReason] || 'end_turn';
}

function buildResponse(statusCode, body) {
  return {
    statusCode,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  };
}

module.exports = { handleAIChat };

File: src/ai_handler.rs
use crate::models::*;
use lambda_http::{Body, Error, Response};
use std::env;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, PartialEq)]
pub enum ApiFormat {
    OpenAI,
    Anthropic,
}

/// Detect API format from request body
pub fn detect_format(body: &serde_json::Value) -> Option<ApiFormat> {
    let has_messages = body.get("messages").and_then(|v| v.as_array()).is_some();
    if !has_messages {
        return None;
    }
    // Anthropic: has max_tokens
    if body.get("max_tokens").is_some() {
        return Some(ApiFormat::Anthropic);
    }
    // Default: OpenAI
    Some(ApiFormat::OpenAI)
}

/// Handle OpenAI Chat Completion format
pub async fn handle_openai_format(
    body: &serde_json::Value,
) -> Result<Response<Body>, Error> {
    let req: OpenAIRequest = serde_json::from_value(body.clone()).unwrap_or(OpenAIRequest {
        model: None,
        messages: None,
        temperature: None,
        max_tokens: None,
        stream: None,
    });

    if req.stream == Some(true) {
        return json_response(400, &serde_json::json!({
            "error": { "message": "Streaming is not supported in Lambda", "type": "invalid_request_error" }
        }));
    }

    let messages = match req.messages {
        Some(ref m) if !m.is_empty() => m,
        _ => {
            return json_response(400, &serde_json::json!({
                "error": { "message": "messages is required and must be a non-empty array", "type": "invalid_request_error" }
            }));
        }
    };

    let model = req
        .model
        .clone()
        .unwrap_or_else(|| default_model());
    let backend = get_backend_for_model(&model);

    if backend == "anthropic" {
        // Convert OpenAI → Anthropic
        let mut system_prompt = String::new();
        let mut anthropic_msgs: Vec<AnthropicApiMessage> = Vec::new();

        for msg in messages {
            if msg.role == "system" {
                if !system_prompt.is_empty() {
                    system_prompt.push('\n');
                }
                system_prompt.push_str(&msg.content);
            } else {
                anthropic_msgs.push(AnthropicApiMessage {
                    role: msg.role.clone(),
                    content: msg.content.clone(),
                });
            }
        }

        let payload = AnthropicApiPayload {
            model: model.clone(),
            messages: anthropic_msgs,
            system: if system_prompt.is_empty() {
                None
            } else {
                Some(system_prompt)
            },
            max_tokens: req.max_tokens.unwrap_or(1024),
            temperature: req.temperature,
        };

        let result = call_anthropic_api(&payload).await?;

        let text = result
            .content
            .as_ref()
            .and_then(|c| c.first())
            .map(|c| c.text.clone())
            .unwrap_or_default();

        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        let input_tokens = result
            .usage
            .as_ref()
            .and_then(|u| u.input_tokens)
            .unwrap_or(0);
        let output_tokens = result
            .usage
            .as_ref()
            .and_then(|u| u.output_tokens)
            .unwrap_or(0);

        let response = OpenAIResponse {
            id: format!(
                "chatcmpl-{}",
                result.id.unwrap_or_else(|| now.to_string())
            ),
            object: "chat.completion".to_string(),
            created: now,
            model: result.model.unwrap_or(model),
            choices: vec![OpenAIChoice {
                index: 0,
                message: OpenAIMessage {
                    role: "assistant".to_string(),
                    content: text,
                },
                finish_reason: map_stop_reason(
                    result.stop_reason.as_deref().unwrap_or("end_turn"),
                ),
            }],
            usage: OpenAIUsage {
                prompt_tokens: input_tokens,
                completion_tokens: output_tokens,
                total_tokens: input_tokens + output_tokens,
            },
        };

        json_response(200, &serde_json::to_value(response)?)
    } else {
        // Call OpenAI directly
        let payload = OpenAIApiPayload {
            model,
            messages: messages.clone(),
            temperature: req.temperature,
            max_tokens: req.max_tokens,
        };

        let result = call_openai_api(&payload).await?;
        json_response(200, &result)
    }
}

/// Handle Anthropic Messages format
pub async fn handle_anthropic_format(
    body: &serde_json::Value,
) -> Result<Response<Body>, Error> {
    let req: AnthropicRequest =
        serde_json::from_value(body.clone()).unwrap_or(AnthropicRequest {
            model: None,
            messages: None,
            system: None,
            max_tokens: None,
            temperature: None,
            stream: None,
        });

    if req.stream == Some(true) {
        return json_response(400, &serde_json::json!({
            "error": {
                "type": "invalid_request_error",
                "message": "Streaming is not supported in Lambda"
            }
        }));
    }

    let messages = match req.messages {
        Some(ref m) if !m.is_empty() => m,
        _ => {
            return json_response(400, &serde_json::json!({
                "error": {
                    "type": "invalid_request_error",
                    "message": "messages is required and must be a non-empty array"
                }
            }));
        }
    };

    let model = req
        .model
        .clone()
        .unwrap_or_else(|| default_model());
    let backend = get_backend_for_model(&model);

    if backend == "openai" {
        // Convert Anthropic → OpenAI
        let mut openai_msgs: Vec<OpenAIMessage> = Vec::new();
        if let Some(ref sys) = req.system {
            openai_msgs.push(OpenAIMessage {
                role: "system".to_string(),
                content: sys.clone(),
            });
        }
        for msg in messages {
            let content_str = match &msg.content {
                serde_json::Value::String(s) => s.clone(),
                serde_json::Value::Array(arr) => arr
                    .iter()
                    .filter_map(|c| c.get("text").and_then(|t| t.as_str()))
                    .collect::<Vec<_>>()
                    .join(""),
                _ => String::new(),
            };
            openai_msgs.push(OpenAIMessage {
                role: msg.role.clone(),
                content: content_str,
            });
        }

        let payload = OpenAIApiPayload {
            model: model.clone(),
            messages: openai_msgs,
            temperature: req.temperature,
            max_tokens: Some(req.max_tokens.unwrap_or(1024)),
        };

        let result = call_openai_api(&payload).await?;

        let choice_text = result["choices"][0]["message"]["content"]
            .as_str()
            .unwrap_or("")
            .to_string();
        let finish_reason = result["choices"][0]["finish_reason"]
            .as_str()
            .unwrap_or("stop");

        let resp = serde_json::json!({
            "id": result["id"].as_str().unwrap_or(&format!("msg_{}", now_secs())),
            "type": "message",
            "role": "assistant",
            "content": [{ "type": "text", "text": choice_text }],
            "model": result["model"].as_str().unwrap_or(&model),
            "stop_reason": map_finish_reason(finish_reason),
            "usage": {
                "input_tokens": result["usage"]["prompt_tokens"].as_u64().unwrap_or(0),
                "output_tokens": result["usage"]["completion_tokens"].as_u64().unwrap_or(0),
            }
        });

        json_response(200, &resp)
    } else {
        // Call Anthropic directly
        let anthropic_msgs: Vec<AnthropicApiMessage> = messages
            .iter()
            .map(|m| {
                let content_str = match &m.content {
                    serde_json::Value::String(s) => s.clone(),
                    serde_json::Value::Array(arr) => arr
                        .iter()
                        .filter_map(|c| c.get("text").and_then(|t| t.as_str()))
                        .collect::<Vec<_>>()
                        .join(""),
                    _ => String::new(),
                };
                AnthropicApiMessage {
                    role: m.role.clone(),
                    content: content_str,
                }
            })
            .collect();

        let payload = AnthropicApiPayload {
            model,
            messages: anthropic_msgs,
            system: req.system,
            max_tokens: req.max_tokens.unwrap_or(1024),
            temperature: req.temperature,
        };

        let result = call_anthropic_api(&payload).await?;
        json_response(200, &serde_json::to_value(result)?)
    }
}

/// Call Anthropic Messages API
async fn call_anthropic_api(
    payload: &AnthropicApiPayload,
) -> Result<AnthropicResponse, Error> {
    let api_key = env::var("ANTHROPIC_API_KEY")
        .map_err(|_| "ANTHROPIC_API_KEY environment variable is not set")?;

    let client = reqwest::Client::new();
    let resp = client
        .post("https://api.anthropic.com/v1/messages")
        .header("Content-Type", "application/json")
        .header("x-api-key", &api_key)
        .header("anthropic-version", "2023-06-01")
        .json(payload)
        .timeout(std::time::Duration::from_secs(28))
        .send()
        .await
        .map_err(|e| format!("Anthropic API request failed: {e}"))?;

    let status = resp.status();
    let text = resp
        .text()
        .await
        .map_err(|e| format!("Failed to read Anthropic response: {e}"))?;

    if !status.is_success() {
        return Err(format!("Anthropic API error ({}): {}", status, text).into());
    }

    serde_json::from_str(&text)
        .map_err(|e| format!("Failed to parse Anthropic response: {e}").into())
}

/// Call OpenAI Chat Completions API
async fn call_openai_api(
    payload: &OpenAIApiPayload,
) -> Result<serde_json::Value, Error> {
    let api_key = env::var("OPENAI_API_KEY")
        .map_err(|_| "OPENAI_API_KEY environment variable is not set")?;

    let client = reqwest::Client::new();
    let resp = client
        .post("https://api.openai.com/v1/chat/completions")
        .header("Content-Type", "application/json")
        .header("Authorization", format!("Bearer {api_key}"))
        .json(payload)
        .timeout(std::time::Duration::from_secs(28))
        .send()
        .await
        .map_err(|e| format!("OpenAI API request failed: {e}"))?;

    let status = resp.status();
    let text = resp
        .text()
        .await
        .map_err(|e| format!("Failed to read OpenAI response: {e}"))?;

    if !status.is_success() {
        return Err(format!("OpenAI API error ({}): {}", status, text).into());
    }

    serde_json::from_str(&text)
        .map_err(|e| format!("Failed to parse OpenAI response: {e}").into())
}

fn get_backend_for_model(model: &str) -> &'static str {
    let lower = model.to_lowercase();
    if lower.starts_with("gpt-") || lower.starts_with("o1") || lower.starts_with("o3") {
        "openai"
    } else {
        "anthropic"
    }
}

fn default_model() -> String {
    env::var("DEFAULT_MODEL").unwrap_or_else(|_| "claude-sonnet-4-20250514".to_string())
}

fn map_stop_reason(reason: &str) -> String {
    match reason {
        "end_turn" => "stop",
        "max_tokens" => "length",
        "stop_sequence" => "stop",
        _ => "stop",
    }
    .to_string()
}

fn map_finish_reason(reason: &str) -> String {
    match reason {
        "stop" => "end_turn",
        "length" => "max_tokens",
        "content_filter" => "end_turn",
        _ => "end_turn",
    }
    .to_string()
}

fn now_secs() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs()
}

fn json_response(
    status: u16,
    body: &serde_json::Value,
) -> Result<Response<Body>, Error> {
    crate::json_response(status, body)
}
